pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerBase.java

```diff
@@ -677,19 +677,16 @@ protected void notifyPendingBatchReceivedCallBack() {
         if (opBatchReceive == null || opBatchReceive.future == null) {
             return;
         }
-        notifyPendingBatchReceivedCallBack(opBatchReceive);
+        synchronized (incomingMessages) {
```

Comment:
can you please explain the reason for this 'synchronized' block ?

Reply:
Looks the 'synchronized' block is to protect the `notifyPendingBatchReceivedCallBack(opBatchReceive)` because of the here https://github.com/apache/pulsar/pull/8207/files/ac48262d4fa431de0364391cc585c9694fefd581..675b28d1755f8968ec51032f80543023d523465d#diff-bbd4739897f6a38e6b966ec9f8795ca5L686 is not a thread-safe operation.

Reply:
Not every call needs to be locked, so I reduced the scope of the lock. Do you think using ReentrantLock is better than using synchronized?

Reply:
Looks it's better to let the MessagesImpl.add() keep thread-safe. So that we can release the lock faster.

Reply:
> can you please explain the reason for this 'synchronized' block ?

The `notifyPendingBatchReceivedCallBack(OpBatchReceive<T> opBatchReceive)` called in `notifyPendingBatchReceivedCallBack()` is not thread-safe, so locking is necessary.

But not every call to `notifyPendingBatchReceivedCallBack()` needs to be locked, only when there is an opBatchReceive that meets the conditions.

The scenarios where thread is not safe are as follows:
Thread A calls incomingMessages.peek(), finds that message A meets the conditions, can join the Messages collection, and is preparing to poll. At this time, thread B polls the message A first. The message polled by thread A is actually message B, but message B does not meet the conditions for joining the Messages set of thread A.

Look forward to your advice.

Reply:
`incomingMessages` is already a concurrent data structure and it does not need this `synchronized` block

Also if you use `synchronized` once you must guard it always with this facility, otherwise you will fall into an inconsistent synchronization.
I don't know very well this code but it looks like there is no other instance of `synchronized` over  `incomingMessages`

Reply:
> `incomingMessages` is already a concurrent data structure and it does not need this `synchronized` block
> 
> Also if you use `synchronized` once you must guard it always with this facility, otherwise you will fall into an inconsistent synchronization.
> I don't know very well this code but it looks like there is no other instance of `synchronized` over `incomingMessages`

Thanks.
I use `synchronized (incomingMessages)`, just to allow synchronized to have a final object that can be locked, because I see no other place to use this object to compete for lock.

I reflected on it, this code can be very confusing, it is better to use `ReentrantLock`.